src/tools/db-upgrade/DbUpgrader.cpp

```diff
@@ -0,0 +1,461 @@
+            tagIndexes_[tagIndex->schema_id.tag_id].emplace(tagIndex);
+        }
+    }
+
+    // Get all edge in space
+    auto edges = schemaMan_->getAllVerEdgeSchema(spaceId_);
+    if (!edges.ok()) {
+        LOG(INFO) << "space Id " << spaceId_ << " no found";
+        return edges.status();
+    }
+    edgeSchemas_ = std::move(edges).value();
+
+    // Get all edge index in space
+    std::vector<std::shared_ptr<nebula::meta::cpp2::IndexItem>> edgeIndexes;
+    iRet = indexMan_->getEdgeIndexes(spaceId_);
+    if (iRet.ok()) {
+       edgeIndexes = std::move(iRet).value();
+    }
+    // handle index
+    for (auto& edgeIndex : edgeIndexes) {
+        if (edgeIndex->schema_id.edge_type > 0) {
+            edgeIndexes_[edgeIndex->schema_id.edge_type].emplace(edgeIndex);
+        }
+    }
+    return  Status::OK();
+}
+
+
+bool DbUpgrader::isValidVidLen(VertexID srcVId, VertexID dstVId) {
+    if (!NebulaKeyUtils::isValidVidLen(spaceVidLen_, srcVId, dstVId)) {
+        LOG(INFO) << "vertex id length is illegal, expect: " << spaceVidLen_
+                  << " result: " << srcVId << " " << dstVId;
+        return false;
+    }
+    return true;
+}
+
+// Used for vertex and edge
+std::string DbUpgrader::encodeRowVal(const RowReader* reader,
+                                     const meta::NebulaSchemaProvider* schema) {
+    auto oldSchema = reader->getSchema();
+    if (oldSchema == nullptr) {
+        LOG(INFO)  << "schema not found.";
+        return "";
+    }
+    std::vector<Value> props;
+    auto iter = oldSchema->begin();
+    size_t idx = 0;
+    while (iter) {
+        auto value = reader->getValueByIndex(idx);
+        if (value != Value::kNullUnknownProp && value != Value::kNullBadType) {
+            props.emplace_back(value);
+        } else {
+            LOG(INFO)  << "Data is illegal.";
+            return "";
+        }
+        ++iter;
+        ++idx;
+    }
+
+    // encode v2 value, use new schema
+    WriteResult wRet;
+    RowWriterV2 rowWrite(schema);
+    for (size_t i = 0; i < props.size(); i++) {
+        wRet = rowWrite.setValue(i, props[i]);
+        if (wRet != WriteResult::SUCCEEDED) {
+            LOG(INFO)  << "Write rowWriterV2 failed";
+            return "";
+        }
+    }
+    wRet = rowWrite.finish();
+    if (wRet != WriteResult::SUCCEEDED) {
+        LOG(INFO)  << "Write rowWriterV2 failed";
+        return "";
+    }
+
+    return std::move(rowWrite).moveEncodedStr();
+}
+
+// Think that the old and new values are exactly the same, so use one reader
+void DbUpgrader::encodeVertexValue(PartitionID partId,
+                                 RowReader* reader,
+                                 const meta::NebulaSchemaProvider* schema,
+                                 std::string& newkey,
+                                 VertexID& strVid,
+                                 TagID   tagId,
+                                 std::vector<kvstore::KV>& data) {
+    if (reader == nullptr) {
```

Comment:
It has been checked before in `doProcessAllTagsAndEdges`

Reply:
Yeah，good ponit！